src/mr_allocator.rs

```diff
@@ -1,60 +1,603 @@
+    /// The correspondence between extent metadata and `raw_mr`
+    #[derive(Debug)]
+    pub(crate) static ref EXTENT_TOKEN_MAP: Arc<ParkingLotMutex<BTreeMap<usize, Item>>> = Arc::new(ParkingLotMutex::new(BTreeMap::<usize, Item>::new()));
+    /// The correspondence between `arena_ind` and `ProtectionDomain`
+    pub(crate) static ref ARENA_PD_MAP: Arc<LockFreeCuckooHash<u32, Arc<ProtectionDomain>>> = Arc::new(LockFreeCuckooHash::new());
+}
+
+/// Combination between extent metadata and `raw_mr`
+#[derive(Debug)]
+pub(crate) struct Item {
+    /// addr of an extent(memory region)
+    addr: usize,
+    /// length of an extent(memory region)
+    len: usize,
+    /// Reference of RawMemoryRegion
+    raw_mr: Arc<RawMemoryRegion>,
+    /// arena index of this extent
+    arena_ind: u32,
+}
+
 /// Memory region allocator
 #[derive(Debug)]
 pub(crate) struct MrAllocator {
     /// Protection domain that holds the allocator
-    pd: Arc<ProtectionDomain>,
+    _pd: Arc<ProtectionDomain>,
+    /// Arena index
+    arena_ind: u32,
 }
 
 impl MrAllocator {
     /// Create a new MR allocator
     pub(crate) fn new(pd: Arc<ProtectionDomain>) -> Self {
-        Self { pd }
+        #[allow(clippy::expect_used)]
+        let arena_ind =
+            init_je_statics(Arc::<ProtectionDomain>::clone(&pd)).expect("init je statics failed");
+        Self { _pd: pd, arena_ind }
     }
 
     /// Allocate a MR according to the `layout`
+    #[allow(clippy::as_conversions)]
     pub(crate) fn alloc(self: &Arc<Self>, layout: &Layout) -> io::Result<LocalMr> {
-        let addr = unsafe { alloc(*layout) };
-        let access = ibv_access_flags::IBV_ACCESS_LOCAL_WRITE
-            | ibv_access_flags::IBV_ACCESS_REMOTE_WRITE
-            | ibv_access_flags::IBV_ACCESS_REMOTE_READ
-            | ibv_access_flags::IBV_ACCESS_REMOTE_ATOMIC;
-        let raw_mr = Arc::new(RawMemoryRegion::register_from_pd(
-            &self.pd,
-            addr,
-            layout.size(),
-            access,
-        )?);
-        Ok(LocalMr::new(raw_mr.addr(), raw_mr.length(), raw_mr))
+        self.alloc_from_je(layout).map_or_else(||{
+            Err(io::Error::new(io::ErrorKind::OutOfMemory, "insufficient contiguous memory was available to service the allocation request"))
+        }, |addr|{
+            let raw_mr = self.lookup_raw_mr(addr as usize);
+            Ok(LocalMr::new(addr as usize, layout.size(), raw_mr))
+        })
+    }
+
+    /// Alloc memory for RDMA operations from jemalloc
+    #[allow(clippy::as_conversions)]
+    fn alloc_from_je(&self, layout: &Layout) -> Option<*mut u8> {
+        let addr = unsafe {
+            tikv_jemalloc_sys::mallocx(
+                layout.size(),
+                (MALLOCX_ALIGN(layout.align()) | MALLOCX_ARENA(self.arena_ind.cast())).cast(),
+            )
+        };
+        if addr.is_null() {
+            None
+        } else {
+            Some(addr.cast::<u8>())
+        }
+    }
```

Comment:
**Unsoundness**: The allocator allows users to obtain (maybe) uninitialized `&mut [u8]`.

Solution: Set the flag `MALLOCX_ZERO`, or change the semantics of memory region types.

The latter may provide higher performance.

Reply:
I'm not sure about the meaning of `change the semantics of memory region types`.
[Providing two APIs](https://github.com/datenlord/async-rdma/issues/54#issuecomment-1106180551) may solve this problem?

Reply:
Exactly